src/Microsoft.OData.Core/UriParser/ODataPathInfo.cs

```diff
@@ -44,6 +46,19 @@ public ODataPathInfo(ODataPath odataPath)
                         this.targetEdmType = collectionType.ElementType.Definition;
                     }
                 }
+
+                // If the last segment is an OperationSegment for a bound operation and the targetNavigationSource is null,
+                // We use the targetNavigationSource for the previous segment.
+                // e.g People/My.Function.GetPeopleWithDogs()
+                if (this.targetNavigationSource == null && lastSegment is OperationSegment operationSegment)
+                {
+                    IEdmOperation operation = operationSegment.Operations.FirstOrDefault();
+
+                    if (operation.IsBound & previous != null)
+                    {
+                        this.targetNavigationSource = previous.TargetEdmNavigationSource;

```

Comment:
@KenitoInc Without even getting into whether this is the right fix, `previous.TargetNavigationSource` is null if `previous` is a `KeySegment` - e.g. for the case where the function is bound to an entity. On the other hand, `(previous as KeySegment).NavigationSource` is not null.